Support date/time in log file name

TuEmb opened the pull request:
Hi @estk,

As https://github.com/estk/log4rs/issues/242. I want to create a PR to apply date/time in log file name.
The path in file appender will be:

```
    path: "log/file-{%Y-%m-%d_%H-%M-%S}.log"
```

It has the same format with `chrono::datetime`

Creative-Difficulty commented:
@estk

bconn98 commented:
FYI @Creative-Difficulty @TuEmb I'll help you guys prep this PR for merging.

However, it looks like in working your change you've removed functionality from the library. Can you please rework this maintain the original functionality?

TuEmb commented:
@bconn98,

Thanks for your reply. Could you tell me which part do I need to rework ? I really want to do that features, because I used `log4rs` in my project and want to record date/time in log file.

Is this `examples/sample_config.yml` ?

bconn98 commented:
@TuEmb The previous solution also handles environment variables as part of the path, unless I'm missing it, I don't see that in the new solution.

bconn98 commented:
Please also update the documentation. Most is generated by rustdoc but we also have some manual markdown as well in the docs directory.

TuEmb commented:
@bconn98, I have updated new commits to keep environment variable as part of the path and updated documentation for date & time format path. Please help me review those changes

bconn98 commented on src/append/file.rs:
```diff
@@ -89,26 +90,51 @@ impl FileAppenderBuilder {
+    fn date_time_format<P: AsRef<Path>>(&self, path: P) -> PathBuf {
+        let path_cow = path.as_ref().to_string_lossy();
+        // Locate the start and end of the placeholder
+        if let Some(start) = path_cow.find('{') {
+            if let Some(end) = path_cow.find('}') {
+                // Extract the date format string
+                let date_format = &path_cow[start + 1..end];
```
This won't work in the case where you have both an environment variable and the date unless the date come first. I.e. the example you updated in `Configuration.md` would fail this code because it would try to resolve the environment variable as your date format

TuEmb replied:
@bconn98 you are right. May I extract the env variable before extract date/time ?
The code will change to:

```
    fn date_time_format<P: AsRef<Path>>(&self, path: P) -> PathBuf {
        let path_cow = path.as_ref().to_string_lossy();
        // Extract the environment path first.
        let env_path: PathBuf = expand_env_vars(path_cow).as_ref().into();

        let date_time_path = env_path.to_str().unwrap();
        // Locate the start and end of the placeholder
        if let Some(start) = date_time_path.find('{') {
            if let Some(end) = date_time_path.find('}') {
                // Extract the date format string
                let date_format = &date_time_path[start + 1..end];
```

I tested with this change, this work with:
 - only ENV path
 - only date time format
 - combine ENV and date time format

bconn98 replied:
Yes this change looks better. A couple notes. 
1. You won't need to try and expand environment variables again at the end of the method.
2. What do you think about adding an identifier similar to how we did the ENV vars? That way it's easier visually to the user as well as easier to parse if we in the future want to add another pattern type to paths.

TuEmb replied:
1. Yes, I updated new commit to just return PathBuf instead of parsing the environment variable again.
2. That's great idea. we can easily handle our pattern in the future. But how do you think about:
 - We will release "date/time for log file" following the chrono format first for users (I see many developers need that feature including me)
 - Then, we will implement the new one that we will do the same as ENV vars to parse easier and can be handled by ourself

9-FS commented:
Is this feature still being worked on?

bconn98 commented:
@9-FS It doesnt look like @TuEmb is interested in taking this to closure. Feel empowered to step in and finish it

9-FS commented:
@bconn98 Thank you for your quick response. Unfortunately, setting up proper logging is literally the very first thing I do when I start a new language, so I think immediately contributing to an established project without ever having written **anything** in Rust is above my ability... For the time being, I will probably set up a custom appender, but I will come back to this in a couple of months should I feel confident enough to meaningfully contribute.

TuEmb commented:
@bconn98 sorry, I assumed that you accepted with my implementation using chrono pattern. Could you help me review again with my latest commits ?

bconn98 commented:
My previous review comment has not been completed. We need a pattern around the date '{}' similar to how the environment variables works. This will allow for you to find it regardless of where it falls in the pattern string.

TuEmb commented:
I added a pattern with `$TIME{chrono_format}` for the log file name. It will be checked after `$ENV` (which is using function `expend_env_vars`). how do you think about that @bconn98  ?

9-FS commented:
@TuEmb Sorry for the question, but is there a reason for `$TIME{}`? Looking at the encoder pattern implementation, as a user I would've expected the `{d()}` syntax again.

bconn98 commented:
@TuEmb I'll take a look this evening. 

@9-FS The reason for the requested change is that we support different patterns when generating log file names. Historically, we support either environment variables using ENV{} or iteration via {}. So the new addition of date/time needs a similarly unique key for us to search for that would not negatively impact the existing options and allows for the pattern to exist anywhere in the string without concern for ordering.

bconn98 commented on src/append/file.rs:
```diff
@@ -89,26 +90,56 @@ impl FileAppenderBuilder {
     /// will be resolved. Note that if the variable fails to resolve,
     /// $ENV{name_here} will NOT be replaced in the path.
     pub fn build<P: AsRef<Path>>(self, path: P) -> io::Result<FileAppender> {
```
Update this comment to reference your TIME option as well.

TuEmb replied:
Thanks for you review. I updated this comment including `$ENV{name_here}` and `$TIME{chrono_format}`

bconn98 commented on src/append/file.rs:
```diff
@@ -89,26 +90,56 @@ impl FileAppenderBuilder {
+
+        let date_time_path = env_path.to_str().unwrap();
+        // Locate the start and end of the placeholder
+        if let Some(start) = date_time_path.find("$TIME{") {
```
Store your pattern as a constant so that you can use `.len()` to determine the length instead of hard coding it.

TuEmb replied:
Thanks for you review. I added const variables for TIME pattern:
```
const TIME_PREFIX: &str = "$TIME{";
const TIME_PREFIX_LEN: usize = TIME_PREFIX.len();
const TIME_SUFFIX: char = '}';
const TIME_SUFFIX_LEN: usize = 1;
```

bconn98 reviewed (commented):
Looks much better! Just a few small comments @TuEmb

bconn98 commented on src/append/file.rs:
```diff
@@ -84,31 +90,67 @@ impl FileAppenderBuilder {
         })
     }
+
+    fn date_time_format<P: AsRef<Path>>(&self, path: P) -> PathBuf {
+        let path_cow = path.as_ref().to_string_lossy();
+        // Extract the environment path first
+        let env_path: PathBuf = expand_env_vars(path_cow).as_ref().into();
```
Let's keep these two calls separate from each other.  I.e. Lets expect the env then call date_time_format instead of calling expand inside of dtf.

TuEmb replied:
I updated to seperate ENV and TIME format. beside that, I also supported multiple placeholders with latest commit

estk commented on docs/Configuration.md:
```diff
@@ -131,6 +131,9 @@ my_console_appender:
 
+The _path_ field also supports date/time formats such as `$TIME{chrono_format}`. Refer
+to [chrono format](https://docs.rs/chrono/latest/chrono/format/strftime/index.html) for date and time formatting syntax
+
```
nice doc!

estk commented on src/append/file.rs:
```diff
@@ -84,31 +90,67 @@ impl FileAppenderBuilder {
+            }
+        }
+        PathBuf::from(date_time_path)
+    }
```
can you add some unit tests for this function? It would be great to just verify some edgecases and verify this is working as intended.

TuEmb replied:
I added unit test for this function:
 - valid `$TIME{}` format
 - invalid `$TIME{}` format
 - no `$TIME{}` format
 - multiple `$TIME{}` in the path

estk replied:
@TuEmb can you please add a test for an un-closed delimiter? What should your code do when you don't find a closing delimiter?

For example:
`logs/log-$TIME{an_invalid_time_format_and_no_closing_brace.log`

Also, in your while statement please an upper limit (lets say 10) to the number of $TIME's you try to parse. We will need to indicate that limit in the docs.

TuEmb commented:
@bconn98 @estk can we go close this PR ?

bconn98 commented:
It’s blocked by #367

samsungapore commented:
reading this thread made me sad

bconn98 commented:
This is blocked by the pr to resolve chrono deprecations.

estk reviewed (changes requested):
@TuEmb Sorry for the delay, I added an inline comment that will need to be addressed due to a possible infinite loop. Overall nearly there with this.

estk commented:
oh i see what happened, there's a merge commit, I'll just squash this then fix it in post. Thanks for your help with this.
